src/rpc.rs

```diff
@@ -12,80 +12,125 @@
  ********************************************************************************/
 use crate::UPClientZenoh;
 use async_trait::async_trait;
-use std::{string::ToString, time::Duration};
+use std::{string::ToString, sync::Arc, time::Duration};
 use tracing::error;
-use up_rust::{RpcClient, RpcClientResult, UAttributesError, UMessage, UMessageError, UUri};
+use up_rust::{
+    communication::{CallOptions, RpcClient, ServiceInvocationError, UPayload},
+    LocalUriProvider, UAttributes, UCode, UMessageType, UPayloadFormat, UPriority, UStatus, UUri,
+};
 use zenoh::prelude::r#async::*;
 
+pub struct ZenohRpcClient {
+    transport: Arc<UPClientZenoh>,
+    uri_provider: Arc<dyn LocalUriProvider>,
```

Comment:
This is interesting. I would have assumed (maybe incorrectly) that you'd have implemented `LocalUriProvider` on `UPClientZenoh`, given that holds the `authority_name`.

Am I missing something?

Tagging in @sophokles73 for his thoughts as I had planned to do so for `UPTransportVsomeip`.

Reply:
This is my original thoughts, too. However, in the [L2 API of up-rust](https://github.com/eclipse-uprotocol/up-rust/blob/946d2fa57e3f44df5c19aba994b911653cff6049/src/communication/in_memory_rpc_client.rs#L188), they take both `UPTransport` and `LocalUriProvider` as arguments, so I want to make them more consistent.
Do you think we can still put `LocalUriProvider` inside the `UPClientZenoh` and provide a different API here?

Reply:
Not sure of the intent here. Pinging @sophokles73 to see if he can help clarify for us.

I thought we could just then pass in an Arc<> to `UPTransportZenoh` for both arguments.

WDYT?

Reply:
What I'm concerned about is how users think while using L2 API. They need to initialize `UPTransportZenoh` with `LocalUriProvider`, and then they should put these two into the L2 structure. I'm not quite sure whether it makes sense or not. Although users don't need to have `LocalUriProvider` in `ZenohRpcClient`, they still face the same issues while using other L2 APIs.

Reply:
If `ZenohRpcClient` needs access to local resource URIs, then it will need to have access to a `LocalUriProvider`. If `ZenohTransport` implements both `UTransport` and `LocalUriProvider`, then you can pass in two Arcs referring to the same `ZenohTransport` instance. If not, then you need to provide Arcs to two different things. Having two separate traits simply allows for more flexibility ...

Reply:
Do you think the additional flexibility is needed in this case @evshary? Or should it be baked into `UPTransportZenoh`? I lean toward the latter, but would like to hear your thoughts.

Reply:
TBH, after @sophokles73's comments below, I feel like I want to take the L2 API away from up-transport-zenoh-rust. Then, it will not be an issue anymore.

The remaining question is, if we implement `LocalUriProvider` in `UPTransportZenoh`, users will use the L2 API like this way: `InMemoryRpcClient::new(Arc<UPTransportZenoh>, Arc<UPTransportZenoh>)`. @PLeVasseur How do you think about it?

Reply:
actually, I'd expect them to do
```rust
let transport = Arc::new(UPTransportZenoh::new());
let rpc_client = InMemoryRpcClient::new(transport.clone(), transport);
...
```
What is the alternative?

Reply:
After a second thought, I think the better way will be `UPTransportZenoh::new(..., uri_provider: &dyn LocalUriProvider)`, and the `LocalUriProvider` can be reused in L2 API.

Reply:
Oh, I didn't see @sophokles73's comments before refresh
I mean this way:
```rust
let uri_provider = LocalUriProvider::new();
let transport = Arc::new(UPTransportZenoh::new(&uri_provider));
let rpc_client = InMemoryRpcClient::new(transport, uri_provider);
...
```

Reply:
what's wrong with that?

Reply:
I don't prefer to have the duplicate argument in the function, but it's just my preference.

Reply:
> TBH, after @sophokles73's comments below, I feel like I want to take the L2 API away from up-transport-zenoh-rust. Then, it will not be an issue anymore.

The bespoke implementation you made is more efficient, right? IMHO it'd be good to keep it so long as it works with the InMemoryRpcServer. :thinking: 

> The remaining question is, if we implement `LocalUriProvider` in `UPTransportZenoh`, users will use the L2 API like this way: `InMemoryRpcClient::new(Arc<UPTransportZenoh>, Arc<UPTransportZenoh>)`. @PLeVasseur How do you think about it?

I suppose this would be the 'default' way of going about things for most users, but they could impl LocalUriProvider if they wanted that flexibility.

Reply:
> I don't prefer to have the duplicate argument in the function, but it's just my preference.

FMPOV there is no _duplicate argument_ since they have different type (and responsibility). The fact that you have a struct which implements both, is nothing more than an implementation detail, isn't it?

Reply:
Ok, then majority rule :laughing:. I'll implement the trait in UPTransportZenoh.